scripts/windows/msys2_prepare.bat

```diff
@@ -11,26 +13,101 @@
 @rem    WIN_JDK_PATH
 @rem    WIN_MSYS2_ROOT
 @rem    WIN_OTP_PATH
+@rem Set VCVARSALL to the proper location of vcvarsall.bat to skip autodetection
+@rem Set MSVC_VERSION to the proper value to skip autodetection
 
-SETLOCAL ENABLEEXTENSIONS
-
-rem Set required vars defaults
+:: Set required vars defaults
 
 IF "%ERTS_VERSION%"=="" SET "ERTS_VERSION=9.3"
-IF "%FORCE_STYRENE_REINSTALL%"=="" SET "FORCE_STYRENE_REINSTALL=false"
-IF "%JDK_URL%"=="" SET "JDK_URL=https://download.java.net/java/GA/jdk11/9/GPL/openjdk-11.0.2_windows-x64_bin.zip"
 IF "%OTP_VERSION%"=="" SET "OTP_VERSION=20.3"
 IF "%PLATFORM%"=="" SET "PLATFORM=x64"
+IF "%WIN_MSYS2_ROOT%"=="" FOR /F %%F IN ('where msys2') DO SET "WIN_MSYS2_ROOT=%%~dpF"
+:: default if autodetect failed
+IF "%WIN_MSYS2_ROOT%"=="" SET "WIN_MSYS2_ROOT=C:\tools\msys64"
+
+IF "%FORCE_STYRENE_REINSTALL%"=="" SET "FORCE_STYRENE_REINSTALL=false"
+IF "%JDK_URL%"=="" SET "JDK_URL=https://download.java.net/java/GA/jdk11/9/GPL/openjdk-11.0.2_windows-x64_bin.zip"
 IF "%WIN_JDK_BASEPATH%"=="" SET "WIN_JDK_BASEPATH=C:\Program Files\Java"
 IF "%JAVA_VERSION%"=="" SET "JAVA_VERSION=11.0.2"
-IF "%WIN_JDK_PATH%"=="" SET "WIN_JDK_PATH=C:\Program Files\Java\jdk-%JAVA_VERSION%"
-IF "%WIN_MSYS2_ROOT%"=="" FOR /F %%F IN ('where msys2') DO SET "WIN_MSYS2_ROOT=%%~dpF"
-IF "%WIN_OTP_PATH%"=="" SET "WIN_OTP_PATH=C:\Program Files\erl%ERTS_VERSION%"
+IF "%WIN_JDK_PATH%"=="" SET "WIN_JDK_PATH=%WIN_JDK_BASEPATH%\jdk-%JAVA_VERSION%"
+IF "%ERLANG_HOME%"=="" SET "ERLANG_HOME=%WIN_OTP_PATH%"
+
+:: Setup MSVC env
+@call:log Find and set MSVC_VERSION
+call:MSVC_VERSION
+IF NOT "%ERRORLEVEL%"=="0" exit /b %ERRORLEVEL%
+
+:: Dump detected version
+@echo SET MSVC_VERSION=%MSVC_VERSION%
+
+@call:log Find and execute the VS env preparation script for %PLATFORM%
+call:VCVARSALL
+IF NOT "%ERRORLEVEL%"=="0" exit /b %ERRORLEVEL%
+
+:: Set the paths appropriately avoiding duplicates
+SET "_PATHS=%WIN_MSYS2_ROOT%\mingw64\bin;%WIN_MSYS2_ROOT%\usr\bin;%WIN_MSYS2_ROOT%"
+ECHO."%PATH%" | FINDSTR /V /C:"%_PATHS%">nul && SET "PATH=%_PATHS%;%PATH%"
```

Comment:
This "ugly" mess is to check if we already have changed the PATH in a previous run. Users may run this script multiple times and we don't want to fill the PATH with duplicates. Shockingly it has a limit of 1024 chars. Visual studio preparation script that is executed earlier shamelessly adds dozens of entries already